spec/lucky/memoize_spec.cr

```diff
@@ -0,0 +1,77 @@
+require "../spec_helper"
+
+include ContextHelper
+
+class OnlyMemoize < Lucky::Action
+  @i = 0
+  memoize def long_task : Int32
+    proc = ->{ @i += 1 }
+    proc.call
+  end
+
+  get "/memoize/one" do
+    long_task
+    text long_task.to_s
+  end
+end
+
+class MultiMemoize < Lucky::Action
+  memoize def task_one : String
+    "running task_one"
+  end
+
+  memoize def task_two : Time
+    Time.utc
+  end
+
+  get "/memoize/multi" do
+    task_one
+    task_two
+    text "Still works"
+  end
+end
+
+class MemoizeOnPage < Lucky::Action
+  @seed = 0
+  memoize def big_data : Int32
+    proc = ->{ @seed += 1 }
+    proc.call
+  end
+
+  expose big_data
+
+  get "/memoze/on-page" do
+    render MemoizePage
+  end
+end
+
+class MemoizePage
+  include Lucky::HTMLPage
+
+  needs big_data : Int32
+
+  def render
+    h1 @big_data.to_s
+    h2 @big_data.to_s
+    h3 @big_data.to_s
+    h4 @big_data.to_s
+    h5 @big_data.to_s
+  end
+end
+
+describe "memoizations" do
+  it "only calls the long_task once" do
+    response = OnlyMemoize.new(build_context, params).call
+    response.body.should contain "1"
+  end
+
+  it "has no conflicts when memoizing numerous methods" do
+    response = MultiMemoize.new(build_context, params).call
+    response.body.should contain "Still works"
+  end
+
+  it "can expose and pass memoized method to page" do
+    data = MemoizeOnPage.new(build_context, params).big_data
+    MemoizePage.new(build_context, big_data: data).render.to_s.should contain %(<h5>1</h5>)
+  end
```

Comment:
This spec uses actions and pages, but I think memoization can apply to any object so you could probably add a fake class here that includes memoizable. You could also hook into the instance to clarify what we're testing. I'm finding it hard to explain so I'll try to write a bit of (untested) code. Something like this:

```crystal
private class ObjectWithMemoizedMethods
  include Lucky::Memoizable
  getter times_method_1_called = 0
  getter times_method_2_called = 0
  
  # Also do this for method_2
  memoize def method_1 : String
    @time_method_1_called += 1 
     "method_1"
  end
end

it "memoizes the method" do
  object = ObjectWithMemoizedMethods.new

  # Do same for method_2
  object.method_1.should eq "method_1"
  2.times { object.method_1 }
  object.times_method_1_called.should eq 1
end
```

If this works then it should be usable in pages, actions and anywhere!

Which makes me think...should we add an charm for `Object` that includes this by default so you can memoize outside actions out of the box? No need to block this PR but something we can think about adding in the future

Reply:
ok, I can dig it. 👍